zxlive/mainwindow.py

```diff
@@ -350,7 +356,47 @@ def reduce() -> None:
             cmd = AddRewriteStep(self.active_panel.graph_view, new_graph, self.active_panel.step_view, reduction["text"])
             self.active_panel.undo_stack.push(cmd)
         return reduce
-    
+
+    def create_new_rewrite(self) -> None:
+        dialog = QDialog()
+        self.rewrite_form = QFormLayout(dialog)
+        name = QLineEdit()
+        self.rewrite_form.addRow("Name", name)
+        description = QTextEdit()
+        self.rewrite_form.addRow("Description", description)
+        left_button = QPushButton("Left hand side of the rule")
+        right_button = QPushButton("Right hand side of the rule")
+        self.left_graph = None
+        self.right_graph = None
+        def get_file(self, button, side) -> None:
+            out = import_diagram_dialog(self)
+            if out is not None:
+                button.setText(out.file_path)
+                if side == "left":
+                    self.left_graph = out.g
+                else:
+                    self.right_graph = out.g
+        left_button.clicked.connect(lambda: get_file(self, left_button, "left"))
+        right_button.clicked.connect(lambda: get_file(self, right_button, "right"))
+        self.rewrite_form.addRow(left_button)
+        self.rewrite_form.addRow(right_button)
+        button_box = QDialogButtonBox(QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel)
+        self.rewrite_form.addRow(button_box)
+        def add_rewrite() -> None:
+            if self.left_graph is None or self.right_graph is None:
+                return
+            rewrite = proof_actions.ProofAction.from_dict({
+                "text":name.text(),
+                "tooltip":description.toPlainText(),
+                "matcher": proof_actions.create_custom_matcher(self.left_graph),
+                "rule": proof_actions.create_custom_rule(self.left_graph, self.right_graph),
+                "type": proof_actions.MATCHES_VERTICES,
+            })
+            proof_actions.rewrites.append(rewrite)
+            dialog.accept()
```

Comment:
Qt does weird stuff with closures in a callback. Check that this continues to work when you add multiple rules, and that dialog doesn't suddenly refer to the previously constructed dialog and gives some kind of segfault.

Reply:
I checked with multiple rules. For some reason, the previous rewrite stops working after adding a new rewrite. The matcher works fine but the rule does not. I'll need to investigate but I am not sure if it is related to Qt.

Reply:
I'm pretty sure that is Qt's doing. It doesn't like it if you use a lambda as a callback. If you replace it by named functions, for instance by def'ing a function inside another function and then return'ing that function, that it might fix it. You can see a couple of examples of this in mainwindow.py I believe.

Reply:
This was just some silly bug where we need to call `ProofActionGroup.copy()`. Fixed in https://github.com/Quantomatic/zxlive/pull/74/commits/516d9f08344374cdc19dfa449d002c7de3e41254